### ISSUE TITLE
Preps for Firebase and Emulator

### ISSUE THREAD

**Issue by m00racle** [m00racle/refract_book#5]
What is count as preparations?

- [ ] I want to setup the firebase SDK locally
- [ ] I want to setup Firebase Emulators also 
- [ ] I want when the system detected that my code runs in localhost it will automatically connect to emulator
- [ ] When my system runs locally (localhost) no connection to live Firebase is allowed
- [ ] PROBABLY the app needs to setup auth also in order to set the connection to emulator.

**Comment by m00racle** [comment]
Here is the documentation of the whole Firebase: [Documentation](https://firebase.google.com/docs/build?hl=en&authuser=0)

**Comment by m00racle** [comment]
- [ ] setup 
- [ ] install firebase SDK (npm install firebase)
- [ ] initialize the project (locally) using firebase init

**Comment by m00racle** [comment]
### Question: firebase-tools login
I remember when installing firebase-tools for the first time I also logging in. Now I do install -g firebase-tools again to make sure I got the latest version of firebase-tools, do I need to login again?

### Answer:
When you install `firebase-tools` globally (`-g` flag) for the first time, you are prompted to log in to your Firebase account during the installation process. This login step is necessary to authenticate the Firebase CLI and associate it with your Firebase account.

If you have already logged in to your Firebase account during the initial installation, you should not need to log in again when running `firebase-tools` commands in the future. The authentication information is stored locally, and subsequent commands will use the existing authentication credentials.

Therefore, reinstalling `firebase-tools` using the `-g` flag should not require you to log in again unless you explicitly log out or the authentication information is somehow removed from your system.

**Comment by m00racle** [comment]
> * [x]  setup cloud function
> * [x]  setup cloud storage
> * [x]  do I need to setup hosting now? --> later on I guess.
> * [ ]  install firebase SDK (npm install firebase)
> * [ ]  initialize the project (locally) using firebase init

**Comment by m00racle** [comment]
I already done these steps on firebase console:

- Create project in Firebase Console
- add app (web app) on the project
- build authentication 
- build firestore database with production rules
- build cloud functions
- build cloud storage and set the server the same as firestore database.
- change my billing and plan to Blaze package

Then I also prepare these local IDE:

- initialize Next.js
- prepare pages and components to deal with frontend UI

Here are things  want to do next:

- Integrate firebase to the local project
- use firebase init to prepare the development environemnt but I just want step by step settings not flag:

Here are things I want to have in local development envirnemt:

- authentication
- firestore
- cloud function
- cloud storage
- emulators

Hosting will be added later.

How do I setup these?

**Comment by m00racle** [comment]
To integrate Firebase with your local project and set up the development environment, you can follow these step-by-step instructions:

1. Install Firebase CLI: Ensure that you have the Firebase CLI installed globally on your machine. You can install it by running the following command:

   ```
   npm install -g firebase-tools
   ```

2. Authenticate with Firebase: Open your terminal and run the following command to log in to your Firebase account:

   ```
   firebase login
   ```

   This will open a browser window where you can log in with your Firebase account credentials.

3. Initialize Firebase in your project directory: Navigate to your project directory (where your Next.js project is located) using the terminal and run the following command:

   ```
   firebase init
   ```

   This command will initialize Firebase in your project and prompt you to select the Firebase features you want to set up. Use the arrow keys to navigate and press the spacebar to select the desired features. In your case, select the following features:

   - Authentication (set up Firebase Authentication)
   - Firestore (set up Firebase Firestore)
   - Functions (set up Firebase Cloud Functions)
   - Storage (set up Firebase Cloud Storage)

   After selecting the features, press Enter to continue.

4. Configure Firebase features: For each selected feature, you'll be prompted to configure specific settings. Follow the prompts and provide the necessary information to set up each feature. For example, you'll need to choose a Firestore database region, set up Firestore security rules, etc. 

   The Firebase CLI will generate the required configuration files and set up the selected Firebase features in your project directory.

5. Set up Emulators (optional): If you want to use Firebase Emulators for local testing, follow the prompts to set up the emulators for each feature you selected. Emulators allow you to simulate Firebase services locally.

   Note that the emulators require additional setup and configuration. You can refer to the Firebase documentation for detailed instructions on setting up and using Firebase Emulators.

6. Start the Emulators (optional): If you have set up the emulators, you can start them by running the following command in your project directory:

   ```
   firebase emulators:start
   ```

   This will start the emulators for the selected Firebase features, allowing you to test and develop locally.

That's it! With these steps, you should have Firebase integrated into your local development environment, including Authentication, Firestore, Cloud Functions, and Cloud Storage. You can now start using Firebase services in your Next.js project and test them locally using the emulators if you have set them up.

Remember to consult the Firebase documentation for more detailed information on specific features and their configuration options.

**Comment by m00racle** [comment]
### Question: npm firebase
Do I need to install in my project directory :
`$ npm firebase` ?

**Comment by m00racle** [comment]
Yes, you will need to install the `firebase` package in your project directory if you haven't done so already. The `firebase` package is the official JavaScript SDK for Firebase and provides the client-side APIs for interacting with Firebase services in your application.

To install the `firebase` package in your project, navigate to your project directory in the terminal and run the following command:

```
npm install firebase
```

This will download and install the `firebase` package and its dependencies in your project's `node_modules` directory. After installation, you can import and use the `firebase` package in your Next.js project to interact with Firebase services such as Authentication, Firestore, Storage, etc.

Make sure to import the required Firebase modules into your code as needed. For example, if you want to use Firebase Authentication, you can import it like this:

```javascript
import firebase from 'firebase/app';
import 'firebase/auth';
```

This will import the `firebase` package and the `auth` module specifically, allowing you to use Firebase Authentication in your application.

Remember to also initialize the Firebase SDK in your project by calling `firebase.initializeApp(config)` with your Firebase project configuration. You can find your project's configuration details in the Firebase Console under the settings of your web app.

With the `firebase` package installed and initialized, you can start using Firebase services in your Next.js project.

**Comment by m00racle** [comment]
Do I install the local firebase SDK before or after I do `$ firebase init` process?

**Comment by m00racle** [comment]
You should install the `firebase` package locally using `$ npm install firebase` before running the `$ firebase init` command.

The `firebase` package is the JavaScript SDK for Firebase, and it provides the client-side APIs for interacting with Firebase services in your application. By installing the `firebase` package locally, you'll have the necessary dependencies available in your project to use Firebase services programmatically.

Once you have installed the `firebase` package, you can proceed with the Firebase CLI initialization process by running `$ firebase init`. The initialization process configures your local project to work with Firebase services and sets up necessary files and configurations. During the initialization, you'll be prompted to choose which Firebase services you want to set up for your project, such as Authentication, Firestore, Storage, and emulators.

So, to summarize:

1. Install the `firebase` package locally in your project by running `$ npm install firebase`.
2. Once the package is installed, you can proceed with the `$ firebase init` process to initialize your project and set up Firebase services.

By following this order, you ensure that you have the required `firebase` package installed in your project before setting up Firebase services with the Firebase CLI.

**Comment by m00racle** [comment]
### Question: 
If I don't set hosting in Firebase can I still use emulators to run the UI? Or I need to just use the npm run dev since I am using Next.js?

**Comment by m00racle** [comment]
Yes, even if you don't set up hosting in Firebase, you can still use emulators to run the UI locally. The emulators allow you to simulate Firebase services locally, including Authentication, Firestore, and Storage.

Since you're using Next.js, you can use the `npm run dev` command provided by Next.js to start your local development server and run your UI. This will provide hot-reloading and other development features provided by Next.js.

To use emulators, you need to run the Firebase emulators separately alongside your Next.js development server. You can start the emulators by running the following command in your project directory:

```bash
$ firebase emulators:start
```

This command will start the Firebase emulators, and you'll be able to access the emulated Firebase services locally. Your Next.js application can interact with the emulated services just like it would with the actual Firebase services.

So, to summarize:

1. Use `npm run dev` to start your Next.js development server for running the UI.
2. Use `firebase emulators:start` to start the Firebase emulators for emulating Firebase services locally.
3. Your Next.js application can interact with the emulated services during development.

By using this approach, you can develop and test your Next.js application locally with the emulated Firebase services, even without setting up Firebase hosting.

**Comment by m00racle** [comment]
### Question: localhost automatically runs in emulator
Can I make the run to be when the app is running in the locahost (ie. using npm run dev) then it will automatically use the emulators for authentication, firestore, cloud function, and cloud storage?

**Comment by m00racle** [comment]
If you want to conditionally use the Firebase emulator settings when running your app on `localhost` and use the real Firebase configuration for other environments, you can modify your Firebase configuration based on the current location or environment.

Here's an example of how you can achieve this in your `apps.js` file:

```javascript
import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/functions';
import 'firebase/storage';

// Your real Firebase configuration with API key and other details
const realFirebaseConfig = {
  // Real Firebase configuration goes here
  // ...
};

// Emulator configuration for authentication, Firestore, functions, and storage
const emulatorConfig = {
  // Emulator configuration goes here
  // Replace the host and port values with the ones specified by the emulators
  authDomain: 'localhost:9099',
  databaseURL: 'http://localhost:8080',
  functionsURL: 'http://localhost:5001',
  storageBucket: 'localhost:9199',
};

// Function to determine if the app is running on localhost
const isLocalhost = () => {
  return window.location.hostname === 'localhost' || window.location.hostname === '127.0.0.1';
};

// Get the Firebase configuration based on the current environment
const getFirebaseConfig = () => {
  return isLocalhost() ? emulatorConfig : realFirebaseConfig;
};

// Initialize Firebase with the appropriate configuration
if (!firebase.apps.length) {
  firebase.initializeApp(getFirebaseConfig());
}
```

In this example, we have separate configurations for the real Firebase settings and the emulator settings. The `getFirebaseConfig` function checks if the app is running on `localhost` based on the `window.location.hostname`. If it is running on `localhost`, it returns the emulator configuration; otherwise, it returns the real Firebase configuration.

By using this approach, you can keep your real Firebase configuration separate in `apps.js` and conditionally use the emulator settings when running the app on `localhost`.

**Comment by m00racle** [comment]
Here is the link to [Emulators Codelab](https://github.com/firebase/emulators-codelab)

**Comment by m00racle** [comment]
Here's the documentation for the provided code snippet:

```javascript
const { el, mount } = redom;

import {
  ItemCardList,
  HeaderIcon,
  HeaderBar,
  ModalDialog,
  CartList
} from "./view.js";

export async function onDocumentReady(firebaseApp) {
  // Print the Firebase configuration options
  console.log("Firebase Config", JSON.stringify(firebaseApp.options));

  const auth = firebaseApp.auth();
  const db = firebaseApp.firestore();

  // Set up emulator settings if running on localhost
  if (location.hostname === "127.0.0.1") {
    console.log("127.0.0.1 detected!");
    auth.useEmulator("http://127.0.0.1:9099");
    db.useEmulator("127.0.0.1", 8080);
  }

  // Create an instance of the HomePage class and mount it to the document body
  const homePage = new HomePage(db, auth);
  mount(document.body, homePage);
}

class HomePage {
  db;
  auth;

  headerBar;
  itemCardList;
  modalDialog;

  cartItems = {};
  cartItemsUnsub;

  constructor(db, auth) {
    this.db = db;
    this.auth = auth;

    // Create instances of HeaderBar, ItemCardList, and ModalDialog
    this.headerBar = new HeaderBar([
      new HeaderIcon("sign_in", "account_circle", "Sign In", () => {
        this.onSignInClicked();
      }),
      new HeaderIcon("cart", "shopping_cart", "N/A", () => {
        this.showCart();
      })
    ]);

    this.itemCardList = new ItemCardList(async (id, data) => {
      try {
        await this.addToCart(id, data);
      } catch (e) {
        console.warn(e);
        this.showError("Error adding item to cart");
      }
    });

    this.modalDialog = new ModalDialog("Cart", "Nothing here.");

    // Create the main element for the page
    this.el = el("div.header-page", [
      this.headerBar,
      this.itemCardList,
      this.modalDialog
    ]);

    // Set up listeners for authentication and items
    this.listenForAuth();
    this.listenForItems();
  }

  listenForAuth() {
    // Listen for changes in authentication state
    this.auth.onAuthStateChanged(user => {
      console.log(`auth.currentUser = ${JSON.stringify(user)}`);
      const signedIn = user !== null;
      this.setSignedIn(signedIn);
    });
  }

  listenForItems() {
    // Listen for changes in the "items" collection
    this.db.collection("items").onSnapshot(items => {
      if (items.size === 0) {
        console.warn("No items in the database ... did you remember to start the emulators with --import?");
      }

      // Update the item card list with the new items
      this.itemCardList.setItems(items);
    });
  }

  async listenForCart(uid) {
    console.log(`listenForCart(${uid})`);

    // Unsubscribe from the previous cart if there was one
    if (this.cartItemsUnsub) {
      this.cartItemsUnsub();
      this.cartItemsUnsub = null;
    }

    // Create the cart document if it doesn't exist
    const cartRef = this.db.collection("carts").doc(uid);
    await cartRef.set(
      {
        ownerUID: uid
      },
      { merge: true }
    );

    // Listen for updates to the cart
    this.cartUnsub = cartRef.onSnapshot(cart => {
      console.log("cart

", cart.data());

      const total = cart.data().totalPrice || 0;
      const count = cart.data().itemCount || 0;
      this.headerBar.setIconText("cart", `\$${total.toFixed(2)} (${count})`);
    });

    // Listen for updates to cart items
    this.cartItemsUnsub = cartRef.collection("items").onSnapshot(items => {
      this.setCartItems(items);
    });
  }

  onSignInClicked() {
    // Handle the sign-in/sign-out button click event
    if (this.auth.currentUser !== null) {
      this.auth.signOut();
    } else {
      this.auth.signInAnonymously();
    }
  }

  setSignedIn(signedIn) {
    // Set the UI state based on the signed-in status
    if (signedIn) {
      this.headerBar.setIconText("sign_in", "Sign Out");
      this.headerBar.setIconEnabled("cart", true);
      this.listenForCart(this.auth.currentUser.uid);
    } else {
      this.headerBar.setIconText("sign_in", "Sign In");
      this.headerBar.setIconText("cart", "N/A");
      this.headerBar.setIconEnabled("cart", false);
      this.setCartItems(null);
    }
  }

  setCartItems(items) {
    let itemIds;

    if (items) {
      this.cartItems = items.docs.map(doc => doc.data());
      itemIds = items.docs.map(doc => doc.id);
    } else {
      this.cartItems = [];
      itemIds = [];
    }

    // Update the add button state for each item card
    this.itemCardList.getAll().forEach(itemCard => {
      const inCart = itemIds.indexOf(itemCard.id) >= 0;
      itemCard.setAddEnabled(!inCart);
    });
  }

  addToCart(id, itemData) {
    if (this.auth.currentUser === null) {
      this.showError("You must be signed in!");
      return;
    }

    console.log("addToCart", id, JSON.stringify(itemData));
    return this.db
      .collection("carts")
      .doc(this.auth.currentUser.uid)
      .collection("items")
      .doc(id)
      .set(itemData);
  }

  showCart() {
    if (this.auth.currentUser === null) {
      return;
    }

    // Display the cart items in a modal dialog
    const items = this.cartItems.map(doc => `${doc.name} - ${doc.price}`);
    this.modalDialog.setContent(new CartList(items));
    this.modalDialog.show();
  }

  showError(e) {
    alert(e);
  }
}
```

I have added comments explaining the purpose and functionality of each section of the code. The comments provide a brief description of what each part does, making it easier to understand the code's structure and logic.

**Comment by m00racle** [comment]
### Question: location.hostname
is location.hostname is javascript reserved syntax?

**Comment by m00racle** [comment]
### Emulator setup:
=== Emulators Setup
? Which Firebase emulators do you want to set up? Press Space to select emulators, then Enter to 
confirm your choices. Authentication Emulator, Functions Emulator, Firestore Emulator, Storage   
Emulator
? Which port do you want to use for the auth emulator? 9099        
? Which port do you want to use for the functions emulator? 5001   
? Which port do you want to use for the firestore emulator? 8080 
? Which port do you want to use for the storage emulator? 9199
? Would you like to enable the Emulator UI? Yes
? Which port do you want to use for the Emulator UI (leave empty to use any available port)? 
? Would you like to download the emulators now? Yes
i  firestore: downloading cloud-firestore-emulator-v1.18.1.jar...
Progress: ======================================================================> (100% of 65MB) 
i  firestore: Removing outdated emulator files: cloud-firestore-emulator-v1.17.4.jar
i  storage: downloading cloud-storage-rules-runtime-v1.1.3.jar...
Progress: ======================================================================> (100% of 53MB) 

i  Writing configuration info to firebase.json...
i  Writing project information to .firebaserc...

+  Firebase initialization complete!

**Comment by m00racle** [comment]
### REVIEW:
this is the part of the [Codelab Emulator](https://github.com/firebase/emulators-codelab/blob/master/codelab-final-state/public/js/homepage.js):

```js
export async function onDocumentReady(firebaseApp) {
  console.log("Firebase Config", JSON.stringify(firebaseApp.options));

  const auth = firebaseApp.auth();
  const db = firebaseApp.firestore();

  if (location.hostname === "127.0.0.1") {
    console.log("127.0.0.1 detected!");
    auth.useEmulator("http://127.0.0.1:9099");
    db.useEmulator("127.0.0.1", 8080);
  }

  const homePage = new HomePage(db, auth);
  mount(document.body, homePage);
}
```
I need to combine this with the normal setting in [expens tracker GitHub repo](https://firebase.google.com/docs/build?hl=en&authuser=0)

**Comment by m00racle** [comment]
This is the codelab have:
```js
export async function onDocumentReady(firebaseApp) {
  console.log("Firebase Config", JSON.stringify(firebaseApp.options));

  const auth = firebaseApp.auth();
  const db = firebaseApp.firestore();

  if (location.hostname === "127.0.0.1") {
    console.log("127.0.0.1 detected!");
    auth.useEmulator("http://127.0.0.1:9099");
    db.useEmulator("127.0.0.1", 8080);
  }

  const homePage = new HomePage(db, auth);
  mount(document.body, homePage);
}
```

**Comment by m00racle** [comment]
I already initialize firebase (firebase init) already set for auth, firestore, cloud storage, cloud function, and emulators. Rules for firestore, storage already set for now. 

Index on firestore.indexes.json also created.

I want to create folder called firebase which consist of firebase.js which set the firebase.

Now, I want (in the setup) to set if the hosting is in localhost it will run to emulators, and else it will connect to the "real" firebase which config use in my firebase console.

**Comment by m00racle** [comment]
To set up the Firebase configuration based on the hosting environment, you can create a `firebase.js` file in your `firebase` folder with the following code:

```javascript
import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/storage';

const firebaseConfig = {
  // Your Firebase config here
  // ...
};

// Initialize Firebase
if (window.location.hostname === 'localhost') {
  // If running in localhost, use Firebase emulators
  firebase.initializeApp(firebaseConfig);
  firebase.auth().useEmulator('http://localhost:9099');
  firebase.firestore().useEmulator('localhost', 8080);
  firebase.storage().useEmulator('localhost', 9199);
} else {
  // If running in production, use the actual Firebase project
  firebase.initializeApp(firebaseConfig);
}

export default firebase;
```

Make sure to replace `yourFirebaseConfig` with the actual configuration from your Firebase console.

In this code, the `firebase.js` file exports the initialized Firebase instance, which can be imported into other files to interact with Firebase services.

The setup conditionally checks if the current hostname is `localhost`. If it is, the Firebase emulators are used by calling the `useEmulator` method on the Firebase services (auth, firestore, and storage). The emulator URLs (`localhost:9099`, `localhost:8080`, `localhost:9199`) should match the ports where your Firebase emulators are running.

If the hostname is not `localhost`, the code initializes Firebase with the actual configuration.

Now, you can import the `firebase.js` file in your application components or modules to use Firebase services. For example:

```javascript
import firebase from './firebase/firebase';

// Use Firebase services
const auth = firebase.auth();
const firestore = firebase.firestore();
const storage = firebase.storage();

// Example usage
auth.signInWithEmailAndPassword('example@example.com', 'password');
firestore.collection('books').get();
storage.ref('images/image.jpg').put(file);
```

Remember to adjust the import path of the `firebase.js` file based on your project structure.

With this setup, your application will connect to the Firebase emulators when running on localhost and connect to the actual Firebase project when running in production.

Let me know if you have any further questions!

**Comment by m00racle** [comment]
### Follow up: location
can I just use location.hostname rather than window.location.hostname?
Also can you add if the host is on 127.0.0.1 also use the emulator

**Comment by m00racle** [comment]
Yes, you can use `location.hostname` instead of `window.location.hostname` in a browser environment. The `location.hostname` property provides the hostname of the current URL.

To include `127.0.0.1` as a condition for using the Firebase emulators, you can modify the code as follows:

```javascript
import firebase from 'firebase/app';
import 'firebase/auth';
import 'firebase/firestore';
import 'firebase/storage';

const firebaseConfig = {
  // Your Firebase config here
  // ...
};

// Initialize Firebase
const { hostname } = location;
if (hostname === 'localhost' || hostname === '127.0.0.1') {
  // If running on localhost or 127.0.0.1, use Firebase emulators
  firebase.initializeApp(firebaseConfig);
  firebase.auth().useEmulator('http://localhost:9099');
  firebase.firestore().useEmulator('localhost', 8080);
  firebase.storage().useEmulator('localhost', 9199);
} else {
  // If running in production, use the actual Firebase project
  firebase.initializeApp(firebaseConfig);
}

export default firebase;
```

In this updated code, both `localhost` and `127.0.0.1` are included in the condition to determine whether to use the Firebase emulators.

By using `location.hostname`, you can access the hostname directly without the need to reference it through the `window` object.

Remember to adjust the import path of the `firebase.js` file based on your project structure.

Now, your application will connect to the Firebase emulators when running on `localhost` or `127.0.0.1` and connect to the actual Firebase project when running in production.

Let me know if you have any further questions!

**Comment by m00racle** [comment]
The error message suggests that there is a conflict between using SWC and Babel for the `next/font` package in Next.js due to a custom Babel configuration.

To resolve this issue, you can try one of the following solutions:

1. Remove the custom Babel configuration: If you have a custom Babel configuration file (e.g., `.babelrc`, `babel.config.js`), try removing it and let Next.js use its default Babel configuration. This can help avoid the conflict with SWC.

2. Disable SWC: Next.js uses SWC by default for improved performance. However, if you encounter compatibility issues, you can disable SWC and rely solely on Babel. To do this, add the following line to your `next.config.js` file:

   ```javascript
   module.exports = {
     swcMinify: false,
   };
   ```

   This configuration will disable SWC minification and allow Babel to handle the compilation.

Make sure to restart your development server (`npm run dev`) after applying these changes.

If the issue persists, it would be helpful to provide more details about your project's configuration and share any relevant code snippets or files.

**Comment by m00racle** [comment]
### PROBLEM: install react-firebaseui
I try to installthe react-firebaseui library but met with errors:

$ npm install react-firebaseui
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: refract_book@1.0.0
npm ERR! Found: react@18.2.0
npm ERR! node_modules/react
npm ERR!   react@"^18.2.0" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer react@">=15 <=17" from react-firebaseui@6.0.0
npm ERR! node_modules/react-firebaseui
npm ERR!   react-firebaseui@"*" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
The suggestions are to use one of the two flags --force or --legacy-peer-dep.

### Answer:
[Case Study and Solution](https://github.com/m00racle/refract_book/wiki/Case-npm-install)

**Comment by m00racle** [comment]
### PROBLEM: updated emulator setup
Source: 
- [auth emulator](https://firebase.google.com/docs/emulator-suite/connect_auth)
- [firestore emulator](https://firebase.google.com/docs/emulator-suite/connect_firestore)
- [cloud storage emulator](https://firebase.google.com/docs/emulator-suite/connect_storage)

SOLUTION: [Review Firebase Setting](https://github.com/m00racle/refract_book/wiki/Review-firebase-setting)

**Comment by m00racle** [comment]
### PROBLEM: failed to download Google Font
```bash
$ npm run dev

> refract_book@1.0.0 dev
> next dev

- ready started server on 0.0.0.0:3000, url: http://localhost:3000
- info Loaded env from F:\Code\refract_book\.env.local
- event compiled client and server successfully in 376 ms (18 modules)
- wait compiling / (client and server)...
AbortError: The user aborted a request.
    at abort (F:\Code\refract_book\node_modules\next\dist\compiled\node-fetch\index.js:1:65190)
    at EventTarget.abortAndFinalize (F:\Code\refract_book\node_modules\next\dist\compiled\node-fetch\index.js:1:65410)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:737:20)     
    at EventTarget.dispatchEvent (node:internal/event_target:679:26)
    at abortSignal (node:internal/abort_controller:314:10)
    at AbortController.abort (node:internal/abort_controller:344:5)
    at Timeout.<anonymous> (F:\Code\refract_book\node_modules\next\dist\compiled\@next\font\dist\google\fetch-css-from-google-fonts.js:35:55)
    at listOnTimeout (node:internal/timers:569:17)
    at process.processTimers (node:internal/timers:512:7) {
  type: 'aborted'
}
- error Failed to download `Roboto` from Google Fonts. Using fallback font instead.
```
